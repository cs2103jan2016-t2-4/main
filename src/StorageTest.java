import static org.junit.Assert.*;
import org.xml.sax.SAXException;
import javax.xml.parsers.*;
import javax.xml.transform.TransformerException;
import java.io.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.LinkedList;
import java.util.List;

import org.custommonkey.xmlunit.XMLAssert;
import org.custommonkey.xmlunit.XMLUnit;
import org.junit.Test;


public class StorageTest {
	
	private static SimpleDateFormat formatter = new SimpleDateFormat("dd-M-yyyy HH:mm:ss");
	
	@Test
	public void testLoadTasksFloating() throws ParserConfigurationException, SAXException, IOException, ParseException {
		File file = new File("sampleXML.xml");
		List<Task> testList =  Storage.loadTasks(file, "floating");
		List<Task> expectedList = new LinkedList<Task>();
		Task expTask1 = testList.get(0);
		Task testTask1 = new Task( "Find potato", formatter.parse("20-2-2016 12:12:12 "),"OPEN");
		expectedList.add(testTask1);
		
		assertTrue(expTask1.equals(testTask1));

	}
	
	@Test
	public void testSaveTasks() throws ParserConfigurationException, SAXException, IOException, TransformerException
	{
		File inputFile = new File("sampleXML.xml");
		File outputFile = new File("test.xml");
		List<Task> testList =  Storage.loadTasks(inputFile, "floating");
		List<Task> testList2 =  Storage.loadTasks(inputFile, "event");
		List<Task> testList3 =  Storage.loadTasks(inputFile, "deadline");
		testList.addAll(testList2);
		testList.addAll(testList3);
		Storage.saveTasks(outputFile, testList);
		
		XMLUnit.setIgnoreWhitespace(true);
		XMLUnit.setIgnoreComments(true);
		XMLUnit.setIgnoreDiffBetweenTextAndCDATA(true);
		XMLUnit.setNormalizeWhitespace(true);

		FileReader fr1 = new FileReader(inputFile);
		FileReader fr2 = new FileReader(outputFile);
		//XMLUnit.compareXML(fr1, fr2);
		XMLAssert.assertXMLEqual(fr1,fr2);


}
	
	
}









