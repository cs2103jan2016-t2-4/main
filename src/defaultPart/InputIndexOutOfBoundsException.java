package defaultPart;

//@@author A0135766W
public class InputIndexOutOfBoundsException extends Exception {
	private int _index;

	public InputIndexOutOfBoundsException(int index) {
		_index = index;
	}

	public int getIndex() {
		return _index;
	}
}
