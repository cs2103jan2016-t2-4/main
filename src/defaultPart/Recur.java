package defaultPart;

public class Recur {

	public enum TimeUnit {
		DAY, WEEK, MONTH, YEAR
	}

	private TimeUnit _timeUnit;
	private int _frequency = 1;
	private TaskDate _endDate;
	private TaskDate _startDate;

	public Recur() {

	}

	public TimeUnit getTimeUnit() {
		return _timeUnit;
	}

	public void setTimeUnit(TimeUnit timeUnit) {
		_timeUnit = timeUnit;
	}

	public int getFrequency() {
		return _frequency;
	}

	public void setFrequency(int frequency) {
		_frequency = frequency;
	}

	public TaskDate getEndDate() {
		return _endDate;
	}

	public void setEndDate(TaskDate endDate) {
		_endDate = endDate;
	}

	public boolean willRecur() {
		return (this.getNextRecur() != null);
	}

	public TaskDate getNextRecur() {
		TaskDate nextDate = (TaskDate) getStartDate().clone();
		/*
		 * TaskDate today = initializeToday(); getNextRecurAfterToday(nextDate, today);
		 */

		incrementNextDate(nextDate);

		if (nextDateAfterEndDate(nextDate)) {
			return null;
		}
		return nextDate;
	}

	private void incrementNextDate(TaskDate nextDate) {
		switch (_timeUnit) {
			case DAY :
				nextDate.add(TaskDate.DATE, this._frequency);
				break;
			case WEEK :
				nextDate.add(TaskDate.DATE, this._frequency * 7);
				break;
			case MONTH :
				nextDate.add(TaskDate.MONTH, this._frequency);
				break;
			case YEAR :
				nextDate.add(TaskDate.YEAR, this._frequency);
				break;
		}
		nextDate.getTimeInMillis();
	}

	private boolean nextDateAfterEndDate(TaskDate nextDate) {
		return _endDate != null && nextDate.compareTo(_endDate) > 0;
	}

	private void getNextRecurAfterToday(TaskDate nextDate, TaskDate today) {
		while (nextDateBeforeToday(nextDate, today)) {
			incrementNextDate(nextDate);
		}
	}

	private boolean nextDateBeforeToday(TaskDate nextDate, TaskDate today) {
		return nextDate.compareTo(today) < 0;
	}

	private TaskDate initializeToday() {
		// initialize "today" to 00:00am of tomorrow
		TaskDate today = new TaskDate();
		today.set(TaskDate.DATE, today.get(TaskDate.DATE) + 1);
		today.set(TaskDate.HOUR_OF_DAY, 0);
		today.set(TaskDate.MINUTE, 0);
		today.set(TaskDate.SECOND, 0);
		today.set(TaskDate.MILLISECOND, 0);
		today.getTimeInMillis();
		return today;
	}

	@Override
	public String toString() {
		String formattedDate = _endDate == null ? "" : _endDate.toString();
		return _frequency + "" + _timeUnit.name().toLowerCase().charAt(0) + " " + formattedDate;
	}

	public TaskDate getStartDate() {
		return _startDate;
	}

	public void setStartDate(TaskDate startDate) {
		_startDate = startDate;
	}
}
