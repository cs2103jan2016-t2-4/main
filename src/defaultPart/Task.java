package defaultPart;

import java.text.SimpleDateFormat;
import java.util.Calendar;

public class Task {

	private String _description;

	private Calendar _date;
	private Calendar _startTime;
	private Calendar _endTime;

	private Recur _recur;
	private boolean _isCompleted;

	public void setDescription(String description) {
		_description = description;
	}

	public Calendar getDate() {
		return _date;
	}

	public void setDate(Calendar date) {
		_date = date;
	}

	public Calendar getStartTime() {
		return _startTime;
	}

	public void setStartTime(Calendar startTime) {
		_startTime = startTime;
	}

	public Calendar getEndTime() {
		return _endTime;
	}

	public void setEndTime(Calendar endTime) {
		_endTime = endTime;
	}

	public String getDescription() {
		return _description;
	}

	public Recur getRecur() {
		return _recur;
	}

	public void setRecur(Recur recur) {
		_recur = recur;
	}

	public boolean isCompleted() {
		return _isCompleted;
	}

	public void toggleCompleted() {
		_isCompleted = !_isCompleted;
	}

	@Override
	public String toString() {
		SimpleDateFormat dateFormat = new SimpleDateFormat("dd-MM-yyyy");
		String formattedDate = "";
		if (_date != null) {
			formattedDate = " | " + dateFormat.format(_date.getTime());
		}
		String timeString = "";
		if (_startTime != null) {
			SimpleDateFormat timeFormat = new SimpleDateFormat("K.mma");
			timeString = " | " + timeFormat.format(_startTime.getTime());
			if (_endTime != null) {
				timeString += "-" + timeFormat.format(_endTime.getTime());
			}
		}
		String recurString = "";
		if (_recur != null) {
			recurString = " | recur=" + _recur;
		}
		return _description + formattedDate + timeString + recurString;
	}
}
