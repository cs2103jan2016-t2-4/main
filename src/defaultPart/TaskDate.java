package defaultPart;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.GregorianCalendar;

public class TaskDate extends GregorianCalendar {
	private static final SimpleDateFormat dateFormat = new SimpleDateFormat("dd-MM-yy");

	@Override
	public boolean equals(Object taskDate) {
		return get(Calendar.DAY_OF_MONTH) == ((Calendar) taskDate).get(Calendar.DAY_OF_MONTH)
				&& get(Calendar.MONTH) == ((Calendar) taskDate).get(Calendar.MONTH)
				&& get(Calendar.YEAR) == ((Calendar) taskDate).get(Calendar.YEAR);
	}
	
	@Override
	public String toString() {
		return dateFormat.format(this.getTime());
	}
}
