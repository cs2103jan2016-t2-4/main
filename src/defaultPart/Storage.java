package defaultPart;

import org.w3c.dom.*;
import org.xml.sax.SAXException;

import defaultPart.Recur.TimeUnit;

import javax.xml.parsers.*;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerConfigurationException;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;

import java.io.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.LinkedList;
import java.util.List;
import java.util.logging.FileHandler;
import java.util.logging.Handler;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.logging.SimpleFormatter;

public class Storage {

	/* For Logging */
	private static final Logger logger = Logger.getLogger(Storage.class.getName());

	/* Date format used to save/load from XML */
	private SimpleDateFormat formatter = new SimpleDateFormat("dd-M-yyyy HH:mm:ss");

	/* Stores current list of tasks in the program */

	private List<Task> _currentTaskList = new LinkedList<Task>();

	/* Used for CommandType.UNDO */
	private List<Task> _prevTaskList = new LinkedList<Task>();

	/**
	 * Get a copy of the task list
	 * 
	 * @return The current Task list
	 */

	public Storage() {
		try {
			Handler handler = new FileHandler("logs/log.txt");
			handler.setFormatter(new SimpleFormatter());
			logger.addHandler(handler);

		} catch (SecurityException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}

	public List<Task> getTaskList() {
		/* Returns a clone to prevent undesired modification */
		return new LinkedList<Task>(_currentTaskList);
	}

	/**
	 * Returns the task at specified index
	 * 
	 * @param index
	 *            Index of task to get
	 * @return Task at specified index
	 */
	public Task getTask(int index) throws IOException {
		if (!isTaskIndexValid(index)) {
			logger.log(Level.WARNING, "Task index \"{0}\" is invalid", index);
			throw new IOException(String.valueOf(index)); // todo: create a new exception for index error?
		}
		return _currentTaskList.get(index);
	}

	/**
	 * Checks if the task index is within the task list size *
	 * 
	 * @param taskIndex
	 *            Index of task to check
	 * @return True if task index is valid
	 */
	public boolean isTaskIndexValid(int taskIndex) {
		return (taskIndex >= 0 && taskIndex < _currentTaskList.size());
	}

	/**
	 * Remove task from task list at specified index
	 * 
	 * @param taskIndex
	 *            Index of task to remove
	 */
	public void removeTask(int taskIndex) {
		_currentTaskList.remove(taskIndex);
	}

	/**
	 * Replace current task list with previous task list, for the "undo" function
	 */
	public void setPreviousListAsCurrent() {
		_currentTaskList = _prevTaskList;
	}

	/**
	 * "Save-state" for future undo operations.
	 */
	public void setCurrentListAsPrevious() {
		_prevTaskList = new LinkedList<Task>(_currentTaskList);
		// TODO: clone all object fields (TaskDate, Recur)
	}

	/**
	 * Add a task into the current task list
	 * 
	 * @param newTask
	 *            Task to be added to task list
	 */
	public void addToTaskList(Task newTask) {
		Calendar newTaskDate = newTask.getDate();
		for (int i = 0; i < _currentTaskList.size(); i++) {
			Calendar taskDate = _currentTaskList.get(i).getDate();
			if (taskDate == null || (newTaskDate != null && newTaskDate.compareTo(taskDate) <= 0)) {
				_currentTaskList.add(i, newTask);
				return;
			}
		}
		_currentTaskList.add(newTask);
	}

	/**
	 * Save the tasks in current task list into an XML file
	 * 
	 * @param file
	 *            File to be saved
	 */
	public void saveTasks(File file) {

		Document doc = initializeDocBuilder();

		// root element
		Element rootElement = doc.createElement("wuriTasks");
		doc.appendChild(rootElement);

		for (Task taskItem : _currentTaskList) {
			createTasksXML(doc, taskItem, rootElement);
		}

		// Save the XML file in a "pretty" format
		transformAndSaveXML(doc, file);
	}

	/**
	 * Load tasks from an XML file into current task list, if file does not exists, function will not attempt
	 * to load tasks ( usually the case when user starts WURI for the first time )
	 * 
	 * @param file
	 *            File to load from
	 * @throws SAXException
	 *             Error in XML file structure
	 * @throws ParseException
	 *             Error in formatting the date
	 */
	public void loadTasks(File file) throws SAXException, ParseException {

		// First check if the file exists and is not a directory but an actual file
		if (file.isFile() && file.canRead()) {

			// Extracts out the list of task nodes
			NodeList nList = extractListFromDocument(file);

			// Iterates through the list of tasks extracted
			for (int temp = 0; temp < nList.getLength(); temp++) {
				{
					Node taskNode = nList.item(temp);
					if (taskNode.getNodeType() == Node.ELEMENT_NODE) {
						Element taskElement = (Element) taskNode;
						Task newTask = null;
						newTask = importTask(taskElement);
						_currentTaskList.add(newTask);
					}
				}
			}
		}
	}

	/**
	 * Transform the XML file into a properly indented and formatted XML document
	 * 
	 * @param doc
	 *            Doc file with all the XML structures
	 * @param file
	 *            Output file to save the formatted XML document
	 */
	private void transformAndSaveXML(Document doc, File file) {
		TransformerFactory transformerFactory = TransformerFactory.newInstance();
		Transformer transformer = null;
		try {
			transformer = transformerFactory.newTransformer();

		} catch (TransformerConfigurationException ex) {
			// Error in Transformer Configuration
			ex.printStackTrace();
			logger.log(Level.FINE, ex.toString(), ex);
			assert false;
		}

		// Properties of the XML format to save the file in
		transformer.setOutputProperty(OutputKeys.DOCTYPE_PUBLIC, "yes");
		transformer.setOutputProperty(OutputKeys.INDENT, "yes");
		transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "3");

		DOMSource source = new DOMSource(doc);
		StreamResult result = new StreamResult(file);
		try {
			transformer.transform(source, result);
		} catch (TransformerException ex) {
			// Error in transformation process
			ex.printStackTrace();
			logger.log(Level.FINE, ex.toString(), ex);
			assert false;
		}
	}

	/**
	 * Instantiate document builder, a commonly used function in Storage component
	 * 
	 * @return Instantiated new Document class
	 */

	private Document initializeDocBuilder() {
		DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
		Document doc = null;
		try {
			DocumentBuilder builder = factory.newDocumentBuilder();
			doc = builder.newDocument();
		} catch (ParserConfigurationException ex) {

			// Will not occur unless builder object is configured wrongly
			ex.printStackTrace();
			logger.log(Level.FINE, ex.toString(), ex);
			assert false;
		}

		return doc;
	}

	/**
	 * Extract out recur class from task and creates the proper XML structure for saving
	 * 
	 * @param doc
	 *            Document which contains the current XML structure
	 * @param taskItem
	 *            Task to extract out the recur class from
	 * @param parentElement
	 *            Parent element to append the recur XML structure to
	 */
	public void extractRecurrFromTask(Document doc, Task taskItem, Element parentElement) {
		Recur recur = taskItem.getRecur();
		if (recur == null) {
			return;
		}
		Element recurrElement = doc.createElement("recur");
		Element recurTimeUnitElement = doc.createElement("timeUnit");
		Element recurFrequencyElement = doc.createElement("frequency");
		Element recurEndOfRecurrElement = doc.createElement("endOfRecurr");
		recurTimeUnitElement.appendChild(doc.createTextNode(recur.getTimeUnit().toString()));
		recurFrequencyElement.appendChild(doc.createTextNode(Integer.toString(recur.getFrequency())));
		recurEndOfRecurrElement
				.appendChild(doc.createTextNode(formatter.format(recur.getEndDate().getTime())));
		recurrElement.appendChild(recurTimeUnitElement);
		recurrElement.appendChild(recurFrequencyElement);
		recurrElement.appendChild(recurEndOfRecurrElement);
		parentElement.appendChild(recurrElement);
	}

	/**
	 * Create the associated XML structure for a Task class, calls extractRecurrFromTask to handle the recur
	 * portion
	 * 
	 * @param doc
	 *            Document which contains the current XML structure
	 * @param rootElement
	 *            Root element of the XML structure
	 * @param taskItem
	 *            Task to extract out the task details from
	 */
	private void createTasksXML(Document doc, Task taskItem, Element rootElement) {
		Element taskElement = doc.createElement("Task");
		Element descriptionElement = doc.createElement("Description");
		Element dateElement = doc.createElement("Date");
		Element startTimeElement = doc.createElement("StartTime");
		Element endTimeElement = doc.createElement("EndTime");
		Element completedElement = doc.createElement("Completed");

		descriptionElement.appendChild(doc.createTextNode(taskItem.getDescription()));
		dateElement.appendChild(doc.createTextNode(getCalendarString(taskItem.getDate())));
		startTimeElement.appendChild(doc.createTextNode(getCalendarString(taskItem.getStartTime())));
		endTimeElement.appendChild(doc.createTextNode(getCalendarString(taskItem.getEndTime())));
		completedElement.appendChild(doc.createTextNode(taskItem.isCompleted() ? "yes" : "no"));
		taskElement.appendChild(descriptionElement);
		taskElement.appendChild(dateElement);
		taskElement.appendChild(startTimeElement);
		taskElement.appendChild(endTimeElement);
		taskElement.appendChild(completedElement);

		// Handles the recurrence section
		extractRecurrFromTask(doc, taskItem, taskElement);

		rootElement.appendChild(taskElement);
	}

	/**
	 * Get a formatted date string from a calendar object
	 * 
	 * @param calendar
	 *            Calendar object to extract string from
	 * @return Formatted string containing date/time
	 */
	private String getCalendarString(Calendar calendar) {
		return (calendar == null) ? "" : formatter.format(calendar.getTime());
	}

	/**
	 * Extract a NodeList object with all the tasks inside the XML file
	 * 
	 * @param file
	 *            File to extract tasks from
	 * @return NodeList object with all the task elements
	 * 
	 * @throws SAXException
	 *             Error in XML file structure
	 */
	private NodeList extractListFromDocument(File file) throws SAXException {

		// Reading the XML file
		DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
		DocumentBuilder builder;
		NodeList nList = null;
		try {

			builder = factory.newDocumentBuilder();
			Document document = builder.parse(file);
			document.getDocumentElement().normalize();

			// Getting all the tasks in the XML structure for this task type
			nList = document.getElementsByTagName("task");

		} catch (ParserConfigurationException ex) {
			// Error in parser configuration
			ex.printStackTrace();
			assert (false);
			logger.log(Level.FINE, ex.toString(), ex);

		} catch (IOException iex) {
			// Error accessing file
			iex.printStackTrace();
			assert (false);
			logger.log(Level.FINE, iex.toString(), iex);
		}

		return nList;

	}

	/**
	 * Import a Task object from a Element object found in the XML structure
	 * 
	 * @param taskElement
	 *            Element object containing the task details
	 * @return Task object described by the specified Element
	 * @throws ParseException
	 *             Error in parsing different date
	 */
	private Task importTask(Element taskElement) throws ParseException {

		// Create new task with extracted description & extract other attributes
		Task newTask = new Task();
		newTask.setDescription(extractStringFromNode(taskElement, "Description"));
		newTask.setDate(extractDateFromNode(taskElement, "Date"));
		newTask.setStartTime(extractDateFromNode(taskElement, "StartTime"));
		newTask.setEndTime(extractDateFromNode(taskElement, "EndTime"));
		if (extractStringFromNode(taskElement, "Completed").equals("yes")) {
			newTask.toggleCompleted();
		}

		// Handles recurrence portion of the task
		Recur taskRecurr;

		taskRecurr = extractRecurFromXML(taskElement);
		newTask.setRecur(taskRecurr);

		// Ensure that recurring tasks imported will recur
		if (taskElement.getElementsByTagName("recur").getLength() == 0) {
			assert (newTask.getRecur().willRecur());
		}
		return newTask;
	}

	/**
	 * Extract out a recur class from the recur XML structure of a task, called when importing tasks
	 * 
	 * @param taskElement
	 *            Element object containing the task/recur details
	 * @return Recur object in taskElement object, null if task does not recur
	 * @throws ParseException
	 *             Error in formatting the date
	 */
	private Recur extractRecurFromXML(Element taskElement) throws ParseException {
		if (taskElement.getElementsByTagName("recur").getLength() == 0) {
			return null;
		}

		String timeUnit = extractStringFromNode(taskElement, "timeUnit");
		int frequency = Integer.parseInt(extractStringFromNode(taskElement, "frequency"));
		Calendar endOfRecurr = extractDateFromNode(taskElement, "endOfRecurr");
		Recur taskRecurr = new Recur();
		taskRecurr.setTimeUnit(TimeUnit.valueOf(timeUnit));
		taskRecurr.setFrequency(frequency);
		taskRecurr.setEndDate(endOfRecurr);
		return taskRecurr;
	}

	/**
	 * Extract a date from node with specified tag and returns as Calendar object
	 * 
	 * @param taskElement
	 *            Element object containing the task details
	 * @param tag
	 *            Tag to specify which date, e.g. "start", "end'
	 * @return Calendar class object converted from the date
	 * @throws ParseException
	 *             Error in formatting the date
	 */
	private Calendar extractDateFromNode(Element taskElement, String tag) throws ParseException {
		String calendarString = taskElement.getElementsByTagName(tag).item(0).getTextContent();
		if (calendarString == "") {
			return null;
		}
		Calendar calendar = new GregorianCalendar();
		calendar.setTime(formatter.parse(calendarString));
		return calendar;
	}

	/**
	 * Extract a string from node with specified tag
	 * 
	 * @param taskElement
	 *            Element object containing the task details
	 * @param tag
	 *            Tag to specify which attribute, e.g. "description"
	 * @return String inside taskElement with specified tag
	 */
	private String extractStringFromNode(Element taskElement, String tag) {
		Node node = taskElement.getElementsByTagName(tag).item(0);
		return (node == null) ? "" : node.getTextContent();
	}
}