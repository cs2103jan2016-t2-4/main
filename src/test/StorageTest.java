package test;

import static org.junit.Assert.*;

import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.text.ParseException;
import java.util.LinkedList;
import java.util.List;
import java.util.Calendar;

import org.custommonkey.xmlunit.XMLAssert;
import org.custommonkey.xmlunit.XMLUnit;
import org.junit.Test;
import org.xml.sax.SAXException;

import defaultPart.InputIndexOutOfBoundsException;
import defaultPart.Recur;
import defaultPart.Storage;
import defaultPart.Task;
import defaultPart.TaskDate;
import defaultPart.TaskTime;

public class StorageTest {

	/* Location to load/save the expected test results */
	private static final String EXPECTED_FILE_NAME = "test/StorageTest_expected.xml";
	private static final String TEST_FILE_NAME = "test/StorageTest_expected.xml";

	/**
	 * Helper function to create a sample task for testing"
	 * 
	 * @param type
	 *            1-Floating 2-Deadline 3-Event 4-Recurring Event
	 * @return new task Created
	 * @throws ParseException
	 */
	public static Task instantiateTestTask(int type) {

		TaskDate calDate = new TaskDate();
		TaskTime calStart = new TaskTime();
		TaskTime calEnd = new TaskTime();
		TaskDate calRecEnd = new TaskDate();
		try {
			calDate.setDateFromString("20/6/2016");
			calStart.setTimeFromString("10:00AM");
			calEnd.setTimeFromString("12:00PM");
			calRecEnd.setDateFromString("20/8/2016");

		} catch (ParseException e) {

			e.printStackTrace();
		}

		Task newTask = new Task();
		switch (type) {
			case 1 :
				newTask.setDescription("Floating Test case");
				break;
			case 2 :
				newTask.setDescription("Deadline Test case");
				newTask.setDate(calDate);
				newTask.setEndTime(calEnd);
				break;
			case 3 :
				newTask.setDescription("Event Test case");
				newTask.setDate(calDate);
				newTask.setStartTime(calStart);
				newTask.setEndTime(calEnd);
				break;
			case 4 :
				newTask.setDescription("Event Recur Test case");
				newTask.setDate(calDate);
				newTask.setStartTime(calStart);
				newTask.setEndTime(calEnd);
				Recur newRecur = new Recur();
				newRecur.setTimeUnit(Recur.TimeUnit.DAY);
				newRecur.setFrequency(3);
				newRecur.setStartDate(calDate);
				newRecur.setEndDate(calRecEnd);
				newTask.setRecur(newRecur);
				break;

		}
		return newTask;
	}

	/**
	 * Helper function to check if two task lists are equal (contain same tasks)
	 * 
	 * @param expectedTask
	 *            Task to compare
	 * @param actualTask
	 *            Task to compare against
	 * @return True if it is the same, false if it is not
	 */
	public static boolean taskListEquals(List<Task> expectedList, List<Task> actualList) {
		if (expectedList.containsAll(actualList) && actualList.containsAll(expectedList)) {
			return true;
		}
		return false;
	}

	/**
	 * Helper function to check if two tasks are equal
	 * 
	 * @param expectedTask
	 *            Task to compare
	 * @param actualTask
	 *            Task to compare against
	 * @return True if it is the same, false if it is not
	 */
	public static boolean taskEquals(Task expectedTask, Task actualTask) {
		if (expectedTask == null || actualTask == null) {
			return false;
		}
		if (!expectedTask.getDescription().equals(actualTask.getDescription())) {
			return false;
		}
		if (expectedTask.isCompleted() != actualTask.isCompleted()) {
			return false;
		}
		Calendar date1 = expectedTask.getDate();
		Calendar date2 = actualTask.getDate();
		if (date1 == null) {
			return date2 == null;
		} else if (!date1.equals(date2)) {
			return false;
		}
		Recur recur1 = expectedTask.getRecur();
		Recur recur2 = actualTask.getRecur();
		if (recur1 == null) {
			return recur2 == null;
		} else if (!recur1.equals(recur2)) {
			return false;
		}
		return true;
	}

	@Test
	public void testGetTaskList() throws SAXException {

		// Setting up expected Task List for comparison
		Task newTaskFloating = instantiateTestTask(1);
		Task newTaskDeadline = instantiateTestTask(2);
		Task newTaskEvent = instantiateTestTask(3);
		Task newTaskRecurEvent = instantiateTestTask(4);
		List<Task> expectedTaskList = new LinkedList<Task>();
		Storage storage = new Storage(new File(TEST_FILE_NAME));

		// This is to test the expected behavior of this function
		expectedTaskList.add(newTaskFloating);
		storage.addToTaskList(newTaskFloating);
		assert (taskListEquals(expectedTaskList, storage.getTaskList()));

		expectedTaskList.add(newTaskDeadline);
		storage.addToTaskList(newTaskDeadline);
		assert (taskListEquals(expectedTaskList, storage.getTaskList()));

		expectedTaskList.add(newTaskEvent);
		storage.addToTaskList(newTaskEvent);
		assert (taskListEquals(expectedTaskList, storage.getTaskList()));

		expectedTaskList.add(newTaskRecurEvent);
		storage.addToTaskList(newTaskRecurEvent);
		assert (taskListEquals(expectedTaskList, storage.getTaskList()));

	}

	@Test
	public void testGetTask() throws IOException, SAXException, InputIndexOutOfBoundsException {

		// Setting up expected Task for comparison
		Task expectedTask = new Task();
		expectedTask.setDescription("Test case");

		// Setting up actual storage behavior
		Storage storage = new Storage();
		storage.addToTaskList(expectedTask);

		// This is to test the expected behavior of this function
		assertEquals(expectedTask, storage.getTask(0));
	}

	public void testIsTaskIndexValid() throws SAXException {

		// Setting up actual storage behavior
		Storage storage = new Storage();
		Task newTaskFloating = instantiateTestTask(1);
		storage.addToTaskList(newTaskFloating);

		// This is a boundary case for the positive value partition
		assertTrue(storage.isTaskIndexValid(0));
		// This is a boundary case for the negative value partition
		assertFalse(storage.isTaskIndexValid(1));

	}

	@Test
	public void testRemoveTask() throws SAXException {

		// Setting up expected Task List for comparison
		Task newTaskFloating = instantiateTestTask(1);
		Task newTaskDeadline = instantiateTestTask(2);
		Task newTaskEvent = instantiateTestTask(3);
		List<Task> expectedTaskList = new LinkedList<Task>();
		expectedTaskList.add(newTaskFloating);
		expectedTaskList.add(newTaskEvent);

		// Setting up the actual storage behavior
		Storage storage = new Storage(new File(TEST_FILE_NAME));
		storage.addToTaskList(newTaskFloating);
		storage.addToTaskList(newTaskDeadline);
		storage.addToTaskList(newTaskEvent);
		storage.removeTask(0);

		// This is to test the expected behavior of this function
		assert (taskListEquals(expectedTaskList, storage.getTaskList()));
	}

	@Test
	public void testAddToTaskList() throws SAXException {

		// Setting up expected Task List for comparison
		Task newTaskFloating = instantiateTestTask(1);
		Task newTaskDeadline = instantiateTestTask(2);
		Task newTaskEvent = instantiateTestTask(3);
		Task newTaskRecurEvent = instantiateTestTask(4);
		List<Task> expectedTaskList = new LinkedList<Task>();
		Storage storage = new Storage(new File(TEST_FILE_NAME));

		// This is to test the expected behavior of this function
		expectedTaskList.add(newTaskFloating);
		storage.addToTaskList(newTaskFloating);
		assert (taskListEquals(expectedTaskList, storage.getTaskList()));

		expectedTaskList.add(newTaskDeadline);
		storage.addToTaskList(newTaskDeadline);
		assert (taskListEquals(expectedTaskList, storage.getTaskList()));

		expectedTaskList.add(newTaskEvent);
		storage.addToTaskList(newTaskEvent);
		assert (taskListEquals(expectedTaskList, storage.getTaskList()));

		expectedTaskList.add(newTaskRecurEvent);
		storage.addToTaskList(newTaskRecurEvent);
		assert (taskListEquals(expectedTaskList, storage.getTaskList()));

	}

	@Test
	public void testSaveTasks() throws SAXException, IOException, ParseException {

		// Load & Save the tasks from the file to see if it saves correctly
		File expectedFile = new File(EXPECTED_FILE_NAME);
		File actualFile = new File(TEST_FILE_NAME);
		Storage storage = new Storage(actualFile);
		storage.loadTasksFromFile();
		storage.saveTasksToFile();

		// Settings for XML formatting
		XMLUnit.setIgnoreWhitespace(true);
		XMLUnit.setIgnoreComments(true);
		XMLUnit.setIgnoreDiffBetweenTextAndCDATA(true);
		XMLUnit.setNormalizeWhitespace(true);

		// This is to test the expected behavior of this function
		FileReader fr1 = new FileReader(expectedFile);
		FileReader fr2 = new FileReader(actualFile);
		XMLAssert.assertXMLEqual(fr1, fr2);
	}

	@Test
	public void testLoadTasks() throws SAXException, IOException, ParseException {

		// Setting up expected Task List for comparison
		Task newTaskFloating = instantiateTestTask(1);
		Task newTaskDeadline = instantiateTestTask(2);
		Task newTaskEvent = instantiateTestTask(3);
		Task newTaskRecurEvent = instantiateTestTask(4);
		List<Task> expectedTaskList = new LinkedList<Task>();
		expectedTaskList.add(newTaskFloating);
		expectedTaskList.add(newTaskDeadline);
		expectedTaskList.add(newTaskEvent);
		expectedTaskList.add(newTaskRecurEvent);

		// Setting up the actual storage behavior
		Storage storage = new Storage(new File(EXPECTED_FILE_NAME));
		storage.loadTasksFromFile();

		// This is to test the expected behavior of this function
		assert (taskListEquals(expectedTaskList, storage.getTaskList()));

	}

}
