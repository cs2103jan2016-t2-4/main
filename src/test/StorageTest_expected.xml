<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<wuriTasks>
   <Task>
      <Description>Floating Test case</Description>
      <Date/>
      <StartTime/>
      <EndTime/>
      <Completed>no</Completed>
   </Task>
   <Task>
      <Description>Deadline Test case</Description>
      <Date>20-6-2016 00:00:00</Date>
      <StartTime/>
      <EndTime>20-6-2016 12:00:00</EndTime>
      <Completed>no</Completed>
   </Task>
   <Task>
      <Description>Event Test case</Description>
      <Date>20-6-2016 00:00:00</Date>
      <StartTime>20-6-2016 10:00:00</StartTime>
	  <EndTime>20-6-2016 12:00:00</EndTime>
      <Completed>no</Completed>
   </Task>
    <Task>
      <Description>Floating Test case</Description>
      <Date/>
      <StartTime/>
      <EndTime/>
      <Completed>no</Completed>
   </Task>
   <Task>
      <Description>Deadline Test case</Description>
      <Date>20-6-2016 00:00:00</Date>
      <StartTime/>
      <EndTime>20-6-2016 12:00:00</EndTime>
      <Completed>no</Completed>
   </Task>
   <Task>
      <Description>Event Recur Test case</Description>
      <Date>20-6-2016 00:00:00</Date>
      <StartTime>20-6-2016 10:00:00</StartTime>
	  <EndTime>20-6-2016 12:00:00</EndTime>
      <Completed>no</Completed>
	  <recur>
         <timeUnit>DAY</timeUnit>
         <frequency>3</frequency>
         <endOfRecurr>20-8-2016 12:00:00</endOfRecurr>
      </recur>
   </Task>
</wuriTasks>
